### ISSUE TITLE
Feature request : IfcStair modifier Wood/Steel : Option to give treads extra depth

### ISSUE THREAD

**Issue by Gorgious56** [IfcOpenShell/IfcOpenShell#3422]
I'm trying to represent this kind of stairs using the IfcStair modifier

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/25156105/82c281c7-2c8f-4a0f-bdf9-7b2ea437b352)

I've tried with the Wood / Steel type and tweaking the tread depth which looked like it would do the trick but treads can't extend behind the tread that is one step above, which I would need for this particular stair.

So I'm short a few centimeters. Sorry I don't know any of the stairs related jargon. I'm sure this has a name :)

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/25156105/715d69fd-1cd1-4d15-8bc2-8b83cfdd780f)

**Comment by Moult** [comment]
Ah yeah this is very important and common. The Australian standards calls the gap a "Tread Gap" so perhaps we can call it a "Tread Overlap".

@Andrej730 would you mind tackling this one? This applies to both wood / steel as well as monolithic concrete stairs.

For wood and steel it is implemented like so - notice this diagram is a Tread Gap, whereas we more commonly want a Tread Overlap (so a negative overlap of -30mm will give a Tread Gap):

![](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fwww.stepform.com.au%2Fimages%2Fstories%2Fvirtuemart%2Fproduct%2FTYPICAL%2520STAIRWAY%2520TERMINOLOGY.jpg&f=1&nofb=1&ipt=6458b9ef14c9e96cbc445f499e2b1732b04c2e3d98c8e8a2aaaf4418db9a93dc&ipo=images)

In Australian standards at least this overlap is _not_ counted as part of the Going (Tread Run), it is in addition, like so:

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/fc5cdce4-230e-48ff-a9ce-bc64665f9e3c)

Also see how on the right diagram there is a zigzag Z shape? That's how this should work for the Concrete stair type.

In theory there's also a third type where there is a "Nosing" which sticks out (like a nose!) which I guess is not technically concrete but for our purposes would be under the Concrete stair generation type:

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/54fcb324-539b-42c0-9257-bef6267edb3e)

(A nosing will also need another dimension for the depth of the nose)

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/0ebe42a8-7bfd-4db5-b993-8b0f408423dd)

So I'd say there needs to be a Tread Overlap parameter added for both wood/steel and concrete (Z shape), it should support negative values, and then another Nosing Depth parameter only for concrete stairs. If the Nosing Depth > 0 and Tread Overlap > 0 then it creates a nosing.

**Comment by theoryshaw** [comment]
Nice request.
Stairs are beasts in any BIM program. ha.

**Comment by Moult** [comment]
Ah interesting in Australia we also do 2R+G. I wonder what other countries use? I don't want to build that range check in the stair geometry cool, but if there was a generic calculator that either calculated common standard values (e.g. 2R+G) or let the user run their own calculation that would be useful.

I'm not a fan of mixing stair tools with handrail tools with stringer tools. They should be three separate tools with different controls.

**Comment by Moult** [comment]
@Andrej730 as @thorstenlang pointed out the term is "Nosing Length" not "Tread Overlap" btw https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/Pset_StairFlightCommon.htm

**Comment by brunopostle** [comment]
2R+G is standard in the UK and the USA

**Comment by thorstenlang** [comment]
Another limit defined in local code may be maximum NumberOfRiser per StairFlight. As well as mimimum Headroom. I do appreciate the stair calculator in my current proprietary BIM tool, when the time is right I would appreciate one in blenderBim.

**Comment by Moult** [comment]
Indeed, but the number of risers is already a shown value so there is no more work we need to do. It is the due diligence of the modeler to ensure it complies with code.

**Comment by DavidDecherf** [comment]
Hi Dion, 

I was one of your studdent at Zigurat. 
In our compagny we design and make stairs and we have a software named Stair Designer. 
How can i help you on this task?

Beside, i continue to learn Python, but i am still at intermediate level. 
I would take part of the BlenderBIM journey.
I can dedicate one hour a week to start.

Regards, 

David

**Comment by brunopostle** [comment]
The old Homemaker stair generator defined a stair with a 2D path, tread/riser parameters, and an overall FTF dimension.

The path followed the wall and the stair was drawn inset sideways, ie you indicate the position of one stringer and the other stringer is determined by the stair width. This is necessary to support non orthogonal spaces.

This allows newel posts, corner landings or winders to be generated automatically. By working upwards, when the required height is achieved the rest of the path can be filled with a horizontal landing.

This system could be extended to insert a mid landing whenever a flight has more than 12 risers etc..

Some pictures, to prove it can be done:
![20230723_090553](https://github.com/IfcOpenShell/IfcOpenShell/assets/1747315/ae8f081a-2b6f-48cd-b84c-e15909306888)
![20230723_090549](https://github.com/IfcOpenShell/IfcOpenShell/assets/1747315/cf2c0f02-9d09-414c-b384-1f3f7b3dc01d)
![20230723_090544](https://github.com/IfcOpenShell/IfcOpenShell/assets/1747315/e28b9f58-ad6c-4a29-bf5c-81a91a654c90)
